File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/consumer/errors/ConditionalRetryListener.kt
package io.micronaut.kafka.docs.consumer.errors

import io.micronaut.configuration.kafka.annotation.ErrorStrategy
import io.micronaut.configuration.kafka.annotation.ErrorStrategyValue
import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.retry.ConditionalRetryBehaviourHandler
import io.micronaut.configuration.kafka.retry.ConditionalRetryBehaviourHandler.ConditionalRetryBehaviour
import io.micronaut.configuration.kafka.exceptions.KafkaListenerException
import io.micronaut.context.annotation.Requires

@Requires(property = "spec.name", value = "RetryProductListenerTest")
// tag::annotation[]
@KafkaListener(
    value = "myGroup",
    errorStrategy = ErrorStrategy(
        value = ErrorStrategyValue.RETRY_CONDITIONALLY_ON_ERROR
    )
)
class ConditionalRetryListener :
    ConditionalRetryBehaviourHandler {
    override fun conditionalRetryBehaviour(exception: KafkaListenerException): ConditionalRetryBehaviour {
        return if (shouldRetry(exception)) {
            ConditionalRetryBehaviour.RETRY
        } else {
            ConditionalRetryBehaviour.SKIP
        }
    }

    // ...

    // end::annotation[]
    companion object {
        private fun shouldRetry(exception: KafkaListenerException): Boolean {
            return true
        }
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/errors/ConditionalRetryListener.py
from micronaut.configuration.kafka.annotation import ErrorStrategy, ErrorStrategyValue, KafkaListener
from micronaut.configuration.kafka.exceptions import KafkaListenerException
from micronaut.configuration.kafka.retry import ConditionalRetryBehaviourHandler
from micronaut.context.annotation import Requires

ConditionalRetryBehaviour = ConditionalRetryBehaviourHandler.ConditionalRetryBehaviour


@Requires(property="spec.name", value="RetryProductListenerTest")
# tag::annotation[]
@KafkaListener(
    value="myGroup",
    errorStrategy=ErrorStrategy(
        value=ErrorStrategyValue.RETRY_CONDITIONALLY_ON_ERROR
    )
)
class ConditionalRetryListener(ConditionalRetryBehaviourHandler):

    def conditionalRetryBehaviour(self, exception: KafkaListenerException) -> ConditionalRetryBehaviour:
        return ConditionalRetryBehaviour.RETRY if self.should_retry(exception) else ConditionalRetryBehaviour.SKIP

    # ...
# end::annotation[]
    @staticmethod
    def should_retry(exception: KafkaListenerException) -> bool:
        return True
